types/time.go

```diff
@@ -3227,32 +3191,23 @@ func yearNumericTwoDigits(t *CoreTime, input string, ctx map[string]int) (string
 }
 
 func yearNumericNDigits(t *CoreTime, input string, ctx map[string]int, n int) (string, bool) {
-	effectiveCount, effectiveValue := 0, 0
-	for effectiveCount+1 <= n {
-		value, succeed := parseDigits(input, effectiveCount+1)
-		if !succeed {
-			break
-		}
-		effectiveCount++
-		effectiveValue = value
-	}
-	if effectiveCount == 0 {
+	year, step := parseNDigits(input, n)
+	if step <= 0 {
 		return input, false
+	} else if step <= 2 {
+		year = adjustYear(year)
 	}
-	if effectiveCount <= 2 {
-		effectiveValue = adjustYear(effectiveValue)
-	}
-	t.setYear(uint16(effectiveValue))
-	return input[effectiveCount:], true
+	t.setYear(uint16(year))
+	return input[step:], true
 }
 
 func dayOfYearThreeDigits(t *CoreTime, input string, ctx map[string]int) (string, bool) {
-	v, succ := parseDigits(input, 3)
```

Comment:
Seems you have changed the behavior of `dayOfYearThreeDigits`.
Please add some test case that len(input) < 3, and compare the result with mysql.

Reply:
MySQL declares that "%j" should be "Day of year (001..366)". But actually, it accepts a number that is up to three digits[2]. (The same with "%d", "Day of the month, numeric (00..31)" but still accept one digit as the day).
TiDB add parsing function `dayOfYearThreeDigits` for '%j', but actually it doesn't support parsing format with '%j' [3]. So it is not easy to add cases for testing this function by testing StrToDate.
Instead of adding cases for 'dayOfYearThreeDigits', I think I will add some test cases for `parseNDigits`, later.

[1] https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_date-format
[2] Some result for '%j' in MySQL 5.7
```
select date, format, str_to_date(date, format) as str_to_date from a;
+-----------+--------+----------------------------+
| date      | format | str_to_date                |
+-----------+--------+----------------------------+
| 015 2021  | %j %Y  | 2021-01-15 00:00:00.000000 |
| 15 2021   | %j %Y  | 2021-01-15 00:00:00.000000 |
| 9 2021    | %j %Y  | 2021-01-09 00:00:00.000000 |
| 900 2021  | %j %Y  | 2023-06-19 00:00:00.000000 |
| 999 2021  | %j %Y  | 2023-09-26 00:00:00.000000 |
| 1000 2021 | %j %Y  | 2000-04-09 00:00:00.000000 | -- don't know why, need further investigation
| 9999 2021 | %j %Y  | 2011-09-26 00:00:00.000000 | -- don't know why, need further investigation
+-----------+--------+----------------------------+
```
[3]
https://github.com/pingcap/tidb/blob/209334951cd920e5408f67fd60a0b542507526ee/types/time.go#L2752-L2756

Reply:
> Instead of adding cases for 'dayOfYearThreeDigits', I think I will add some test cases for `parseNDigits`

Find it hard to add test cases for the private method `parseNDigits`, add more test cases for `StrToDate` instead.